File: c++/Container With Most Water.cpp
#include <iostream>
#include <vector>
#include <algorithm>

class Solution {
public:
    int maxArea(std::vector<int>& height) {
        int maxWater = 0;
        int left = 0, right = height.size() - 1;

        while (left < right) {
            int h = std::min(height[left], height[right]);
            maxWater = std::max(maxWater, h * (right - left));
            
            if (height[left] < height[right]) {
                left++;
            } else {
                right--;
            }
        }

        return maxWater;
    }
};
//  test case
int main() {
    Solution sol;
    std::vector<int> height = {1, 8, 6, 2, 5, 4, 8, 3, 7};
    int max_water = sol.maxArea(height);
    std::cout << "max water: " << max_water << std::endl;
    return 0;
}

File: python/Container With Most Water.py
from typing import List

class Solution:
    def maxArea(self, height: List[int]) -> int:
        left, right = 0, len(height) - 1
        max_water = 0
        
        while left < right:
            width = right - left
            h = min(height[left], height[right])
            current_water = width * h
            max_water = max(max_water, current_water)
            if height[left] < height[right]:
                left += 1
            else:
                right -= 1
        
        return max_water

# test case
if __name__ == "__main__":
    solution = Solution()
    
    # case 1
    height = [1,8,6,2,5,4,8,3,7]
    result = solution.maxArea(height)
    print(f"Maximum amount of water the container can store: {result}") 
